### ISSUE TITLE
Create YapDatabaseREST: Sync YapDatabase with REST Backend

### ISSUE THREAD

**Issue by iosdev-republicofapps** [yapstudios/YapDatabase#343]
Hi,

I commented on https://github.com/yapstudios/YapDatabase/issues/37 but I'd like to create a YapDatabase extension that synchronizes a local YapDatabase with a remote REST API.  For example a notes app where the data is persisted "for real" on a remote data store fronted by a REST API (e.g. a  Rails app) but is cached locally in YapDatabase to support "offline mode" where reads and writes can be done offline (e.g. on an airplane) and then synced with the remote when online again.

Any suggestions or hints would be appreciated.  How would you suggest I get started?

Here are some initial questions I can think off:

1. What is the ultimate authority on "truth" when it comes to data?  Remote or local?
2. How do we tell "who wins" when there were "concurrent" modifications on both the remote and local?  What is time, local time with skew or server time?
3. How do we handle merge conflicts when the remote and local versions conflict?
4. How should the networking be done?  With AFNetworking, RESTKit or configurable plugins?
5. How with authentication be handled?
6. How will network errors be handled?
7. How opinionated should this be?  Allow minimal or maximal configuration?
8. Can the user "watch" for changes to remote data?
9. How much of the remote data is synced?

For my personal use case, I'd actually prefer to use Firebase but their Mac OS X support seems to be going away.

Here are some of my "suggested" answers to the above based on my personal preferences:

1. What is the ultimate authority on "truth" when it comes to data?  Remote or local?
    * **Answer:** User is given choice of policies: ServerAlwaysWins, LocalAlwaysWins, LatestWinsByTime, LatestWinsByLockVersion.  Default would be ServerAlwaysWins.
2. How do we tell "who wins" when there were "concurrent" modifications on both the remote and local?  What is time, local time with skew or server time?
    * **Answer:** Assume skew is not too large or use optimistic locking?
3. How do we handle merge conflicts when the remote and local versions conflict?
    * **Answer:** No merges - either server or local wins completely.  How do we handle merge conflicts with trees of objects where children have merge conflicts?
4. How should the networking be done?  With AFNetworking, RESTKit or configurable plugins?
    * **Answer:** Maybe RESTKit.
5. How with authentication be handled?
    * **Answer:** Support either username/password or username/token.  Maybe Oauth eventually but it is complicated.
6. How will network errors be handled?
    * **Answer:** Retry on 5xx error and 401 error, else no retry.
7. How opinionated should this be?  Allow minimal or maximal configuration?
    * **Answer:** I favor minimal.  Make it "opinionated" like Rails.
8. Can the user "watch" for changes to remote data?
    * **Answer:** Maybe.  But I don't want to get into sockets.
9. How much of the remote data is synced?
    * **Answer:** Two options: All or AsNeeded.

I'm really doing this for a personal project and releasing it to the public out of the goodness of my heart.  So I want to maximize utility for my project while still maintaining some flexibility for others to use it and making decent design choices.  But I'm **not** looking to make the "be all to end all" of REST sync options - I am a lone hobby developer and don't have time to make a giant solution.  But I want to do it right as much as possible.

Any thoughts?

**Comment by robbiehanson** [comment]
I just released the [YapDatabaseCloudCore](https://github.com/yapstudios/YapDatabase/tree/v3.0/YapDatabase/Extensions/CloudCore) extension in the 3.0 branch. This extension is in beta. But it does a lot of what we're discussing here. And can be used to achieve a REST solution (for a variety of server architectures).

Here are my thoughts:

1. What is the ultimate authority on "truth" when it comes to data? Remote or local?
  - **Answer:** By necessity, the server must provide the final answer on "truth". This is because, in a sync situation, other devices will only see what's on the server. (E.g. deviceB won't see what's on deviceA, it will only see what's in the cloud.)

2. How do we tell "who wins" when there were "concurrent" modifications on both the remote and local? What is time, local time with skew or server time?
  - **Answer:** Time doesn't matter. Only the server matters. The winner is whoever the server says is the winner. (Which is generally the person/device that uploads to the server first.)

    For example, consider syncing to a Dropbox-style system:
When a file/record is changed, that change needs to be uploaded to the cloud server. The client has the previous eTag of the file in the cloud. It then attempts an upload to the server with an "If-Match", specifying the value as the previous eTag. Now, if two devices attempt an upload simultaneously, the server will only allow one to win. The other will have to handle the conflict, and perform the proper merge.

2. How do we handle merge conflicts when the remote and local versions conflict?

  - **Answer:** Merge conflicts are interesting because their implementation can be application specific. However, in the vast majority of cases, the logic is the same for record-based merge conflicts. Here's the algorithm.

    (Keep in mind that most sync services do NOT tell you what changed. They will only give you the most recent version of the record. And the onus is on the client to figure out everything.)

    - The client maintains a list of changes to a record, including the keys for which values have changed, as well as both the previous & current value for those changed keys.
    - With only this information, the client can compare it's version, with the version on the server, and detect, on a per-property basis, if the value was changed only locally, only remotely, or both (in which case conflict resolution must occur).

    Generally, the client will allow the remote to win. That is, if a property was changed remotely and locally, the remote change wins (because it made it to the server first).

3. How should the networking be done? With AFNetworking, RESTKit or configurable plugins?

  - **Answer:** The networking is dependent on the server. For example, both Dropbox & Amazon provide SDK's. Thus, the networking should be independent of the generic REST extension.

    In other words, I believe the best solution is to write a generic "CloudCore" extension that handles the basics. And allow others to supply the networking that's specific for their own server solution.

4. How with authentication be handled?

  - **Answer:** This is part of the networking, and should be pluggable.

5. How will network errors be handled?

  - **Answer:** This is part of the networking, and should be pluggable.

6. How opinionated should this be? Allow minimal or maximal configuration?

  - **Answer:** Maximum flexibility is generally preferred. Otherwise you'll end up with an architecture which only works for a few select cloud providers. And you'll have to start from scratch again if you need to transition to a different cloud solution. (Witness my implementation of YapDatabaseCloudKit, which only works with Apple's CloudKit. And then I had to go and re-write it...)

7. Can the user "watch" for changes to remote data?

  - **Answer:** Definitely. This shouldn't be part of "CloudCore".

8. How much of the remote data is synced?

  - **Answer:** This is app specific. And shouldn't be part of "CloudCore".

**Comment by robbiehanson** [comment]
I've started writing the documentation [here](https://github.com/yapstudios/YapDatabase/wiki/YapDatabaseCloudCore).

**Comment by flovouin** [comment]
Hi,
I stumbled upon this "issue" and am interested in this functionality. I took a look at the code for `YapDatabaseCloudCore` but haven't played with it just yet. It would seem that it's already in a functional state. @robbiehanson Could you please tell me what's the status on this? Is it usable in production? Is there an estimate for the release of YapDatabase 3.0?
Thanks a lot for the hard work on this framework!
Cheers,
Flo

**Comment by robbiehanson** [comment]
> Could you please tell me what's the status on [YapDatabaseCloudCore]? Is it usable in production?

YapDatabaseCloudCore is indeed functional. I've been working on the concept for months now (in my own project), and have been refining it. In fact, I went ahead and pushed (to the v3.0 branch) an [updated revision](https://github.com/yapstudios/YapDatabase/tree/v3.0/YapDatabase/Extensions/CloudCore) of it.

My goal is for YDBCloudCore to ultimately be able to support drastically different sync architectures. Such as:
- Apple's CloudKit
- Dropbox-style file-based stuff
- Custom sync solutions people write for their own server

This is a pretty tall order. But I think it's do-able if we distill only the very basics into YDBCloudCore. Then we make it easy to extend this layer, in order to customize for the specific flavor of sync architecture.

In my first attempt, I went overboard in targeting Dropbox-style file-based systems. And, indeed, when I attempted to use it with my own custom sync server, I ended up fighting with YDBCloudCore to get it to do what I needed. In the end, I went back and "distilled" the solution some more. Basically, I ripped out a bunch of code that was very specific to the Dropbox-style architecture.

Unfortunately, the documentation is lacking. If you're interested in using it, hit me up on skype (robbiehanson15) so I can answer questions.

> Is there an estimate for the release of YapDatabase 3.0?

I'm hoping for November. But I'm currently very busy with a startup. So it will depend on how much time I'm able to carve out for open source work.

**Comment by flovouin** [comment]
Robbie,

Thanks very much for your quick reply and getting into details. My use case is indeed a custom sync solution, so I'm happy to hear that you made it easier to integrate with CloudCore.

I've started playing with the new version you pushed and was able to build a basic proof of concept. So far CloudCore seems to achieve exactly what I'm looking for, but I'd like to ask you a few questions to make sure I'm using it correctly.

- One must subclass `YapDatabaseCloudCore` and implement its `newConnection`, which should it turn provide a subclassed `YapDatabaseCloudCoreConnection`, for which `newRead[Write]Transaction` have also been overriden. Finally (this is the interesting bit) the returned transaction should be a subclass of `YapDatabaseCloudCoreTransaction`, implementing the "`did`" methods, to listen for database changes. In these methods, `addOperation` can be used to queue operations performing server calls. One can then register a `YapDatabaseCloudCorePipeline` with the extension. The pipeline's delegate takes care of performing the operations, and either completes an operation inside a `readWrite` transaction, or delays it if it failed, by setting its status as pending. _Is this workflow correct?_
- Now that there's a single `YapDatabaseCloudCoreOperation` type, what should be stored in the generic `persistentUserInfo` dictionary? Right now I'm only storing the key for the relevant object, such that the pipeline's delegate has to first read back the object from the database before performing a call to the remote service. The object itself contains the list of local changes, much like the `changedProperties` of `MyDatabaseObject`. The consequence of this approach is that the first operation to be queued for a given object will read all changes for this object up to the delegate's call. If several operations for the same object have been queued, this might make the following operations obsolete. This is something that could be solved by calling `skipOperationsPassingTest`, but _I was wondering if the overall approach was the one intended?_ The other solution I see is to use the `persistentUserInfo` dictionary to store more details to fully describe the operation. However it looks to me like I would store a lot of redundant data, both in the operation and in the object itself, to track the atomic changes to its properties.
- _If an operation fails, can I somehow update it (e.g. to keep track of the number of retries, set the exponential backoff, etc.)?_ This is probably what `modifyOperation` does, but just checking.

One last thing: the project I'm working on at the moment is in Swift. I was able to use `CloudCore` by updating `YapDatabase.podspec`. However because I subclassed several `CloudCore` objects I also needed to access private headers. I temporarily commented the `private_header_files` declaration in the `podspec`, but this prevents me from using frameworks, as I would also need to expose them in the `modulemap`. At the moment I use a bridging header, which works just fine. I realise that version 3.0 is a work in progress, so no rush here. I just wanted to get your attention on this if subclassing `CloudCore` components is indeed the correct approach.

Thanks again,

Florent
